### ISSUE TITLE
MonoL1.l1 is hard to tune with Adam-family optimizers

### ISSUE THREAD

**Issue by LarocheC** [calad0i/HGQ2#12]
## Environment

- HGQ 0.1.8
- Keras 3
- Torch backend
- `keras.optimizers.AdamW`

## Summary

Sorry for the two issues back-to-back. I ran into strange behavior when using `MonoL1` to push down the bit-width parameters.

The gradient-based automatic bit-width optimization and `MonoL1` expose an `l1` coefficient that looks like a continuous tuning knob for the strength of the bit-width pressure. However, when the bit-width parameters are optimized with an Adam-family optimizer, the magnitude of `MonoL1.l1` has very little practical effect. Adam's per-parameter normalization largely removes the gradient magnitude that the `l1` coefficient is meant to control.

In practice, at least in my experiments, `MonoL1` behaves more like an on/off switch:

- adding it pushes the bit-widths down,
- but changing `l1` over orders of magnitude does not meaningfully change the bit-width trajectory.

This is easy to miss because training still runs normally.

## Reproduction

Minimal standalone reproduction using the torch backend:

```python
import os
os.environ.setdefault("KERAS_BACKEND", "torch")

import numpy as np
import keras
import torch

from hgq.layers import QDense
from hgq.config import QuantizerConfigScope
from hgq.regularizers import MonoL1

ADAM_LR = 1e-3
SGD_LR = 1e-3
EPS = 1e-8

rng = np.random.default_rng(0)


def bit_grad(lam):
    """Build one QDense with MonoL1(lam), run one forward/backward pass,
    and return the mean absolute gradient on the bit-width parameter."""
    with QuantizerConfigScope(
        q_type="kbi",
        place="weight",
        b0=6,
        i0=0,
        br=MonoL1(lam),
    ):
        layer = QDense(64)

    layer.build((None, 64))

    x = keras.ops.convert_to_tensor(
        rng.standard_normal((32, 64)).astype("float32")
    )

    loss = keras.ops.mean(layer(x, training=True) ** 2) + sum(layer.losses)
    loss.backward()

    bvar = [
        w for w in layer.trainable_weights
        if "quantizer" in w.path and w.path.split("/")[-1] == "b"
    ][0]

    return float(torch.mean(torch.abs(bvar.value.grad)))


for lam in (1e-2, 1e-1):
    g = bit_grad(lam)

    # Adam first step is approximately lr * sign(grad)
    adam_step = ADAM_LR * g / (np.sqrt(g * g) + EPS)

    # Plain SGD keeps the gradient scale
    sgd_step = SGD_LR * g

    print(
        f"lambda={lam:7.0e}  "
        f"|grad|={g:.6e}  "
        f"Adam-step={adam_step:.6e}  "
        f"SGD-step={sgd_step:.6e}"
    )
```

Output:

```text
lambda=  1e-02  |grad|=1.005991e-02  Adam-step=9.999990e-04  SGD-step=1.005991e-05
lambda=  1e-01  |grad|=1.000592e-01  Adam-step=9.999999e-04  SGD-step=1.000592e-04
```

A 10x increase in `l1` yields a 10x larger bit-width gradient, as expected.
With plain SGD, this would also yield a 10x larger update.

With Adam, the first-step update remains essentially unchanged.

## Why this happens

`MonoL1` adds a term of the form

```python
loss += l1 * sum(b)
```

where `b` is the trainable bit-width parameter. Therefore, the gradient of this regularizer with respect to each bit-width parameter is simply:

```python
d loss / d b = l1
```
With plain SGD, that directly changes the update size:

```python
update = lr * grad = lr * l1
```

so `l1` behaves like a linear strength knob.

With Adam-style optimizers, however, the update is normalized by an estimate of the gradient magnitude:

```python
update ≈ lr * grad / sqrt(grad^2)
```

If the gradient is dominated by the constant `MonoL1` term, then:

```python
update ≈ lr * l1 / abs(l1)
       ≈ lr
```

The scale of `l1` is divided out. In that regime, changing `l1` from `1e-2` to `1e-1` changes the raw gradient, but it does not meaningfully change the Adam step size. The regularizer behaves more like an on/off switch than a tunable strength parameter.

This is not specific to the toy example above. In my QAT experiments with `AdamW`, using `l1 = 1e-3`, `1e-2`, or `1e-1` led to effectively identical bit-width trajectories and identical validation error to four decimals.

## Suggested fix

I'm not sure how to properly fix this, but I think even documenting the interaction would already help a lot. Right now as a walkaround, I route the route bit-width parameters to plain SGD. 

Or maybe I'm doing something wrong :-)

**Comment by calad0i** [comment]
No worries for the issues : )

The gradient to the bitwidths come from two sources: the differentiable quantization itself pushing grad towards infinity, and the one from ebops/L1 pushing it towards zero. The important part is the difference between these two, and the overall gradient shall vanish near the optimal. In your case, it seems anything larger than 1e-3 annihilates the other gradient.

p.s., L1=1e-8 is set for the default parameter, actually. If beta is disabled, L1 around ~1/10 the working beta values should be in the right ballpark.

**Comment by LarocheC** [comment]
Thanks for the feedback. After a bit more tuning, I managed to get some results. And sorry for the noise. :-)
